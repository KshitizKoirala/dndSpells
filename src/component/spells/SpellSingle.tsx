import { useState } from "react";

interface SpellProps {
  spellName: string;
}

function SpellSingle({ spellName }: SpellProps) {
  const [isFavourite, setIsFavourite] = useState(false);

  return (
    <li className="spells-single" data-testId="spellsSingleComponent">
      <p>{spellName}</p>
      <button
        type="button"
        className="view-btn"
        onClick={() => setIsFavourite(!isFavourite)}
      >
        <img
          className="favourite-btn"
          src={isFavourite ? "favourite-yes.png" : "favourite.png"}
          alt="favourite"
        />
      </button>
    </li>
  );
}

export default SpellSingle;
