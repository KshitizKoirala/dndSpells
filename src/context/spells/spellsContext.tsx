// import React, { useContext, useState } from "react";

// type ThemeContextProviderProps = {
//   children: React.ReactNode;
// };

// const SpellsContext = React.createContext([]);

// export const useFavourites = () => useContext(SpellsContext);

// export function UserContextProvider({ children }: ThemeContextProviderProps) {
//   // Creating a spells state to be accessed overall
//   const [spells, setSpells] = useState([]);
//   const value = React.useMemo(
//     () => ({
//       spells,
//       setSpells,
//     }),
//     [spells],
//   );

//   return (
//     <SpellsContext.Provider value={value}>{children}</SpellsContext.Provider>
//   );
// }

function spellsContext() {
  return <div>spellsContext</div>;
}

export default spellsContext;
