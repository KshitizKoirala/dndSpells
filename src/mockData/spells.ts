declare namespace SpellType {
  interface Single {
    index: string;
    name: string;
    url: string;
  }
  interface List {
    count: number;
    results: Array<Single>;
  }
}

export const spellData: SpellType.List = {
  count: 5,
  results: [
    {
      index: "acid-arrow",
      name: "Acid Arrow",
      url: "/api/spells/acid-arrow",
    },
    {
      index: "acid-splash",
      name: "Acid Splash",
      url: "/api/spells/acid-splash",
    },
    {
      index: "aid",
      name: "Aid",
      url: "/api/spells/aid",
    },
    {
      index: "alarm",
      name: "Alarm",
      url: "/api/spells/alarm",
    },
    {
      index: "alter-self",
      name: "Alter Self",
      url: "/api/spells/alter-self",
    },
  ],
};

export default spellData;
