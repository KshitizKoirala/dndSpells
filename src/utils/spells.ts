import axios from "axios";
import api from "./api";

/**
 * FETCH ALL SPELLS FROM API
 * @export fetchAllSpells Function
 */

const fetchAllSpells = async () => {
  await axios.get(api.fetch).then((res) => {
    const spells = res.data.results;
    return spells;
  });
};

export default fetchAllSpells;
