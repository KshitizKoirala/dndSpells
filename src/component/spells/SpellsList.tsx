import axios from "axios";
import { useEffect, useState } from "react";

import api from "../../utils/api";
import SpellSingle from "./SpellSingle";

// Type Definition of Spells Array
interface IMapSpell {
  index: string;
  name: string;
}

function SpellsList() {
  const [allSpells, setAllSpells] = useState([]);

  useEffect(() => {
    const getAllSpells = async () => {
      await axios.get(api.fetch).then((res) => {
        setAllSpells(res.data.results);
      });
    };
    getAllSpells();
  }, []);

  return (
    <div className="spells-parent" data-testid="spells-list">
      <ul className="mt-3 spells-list">
        {allSpells.map((spell: IMapSpell) => (
          <SpellSingle key={spell.index} spellName={spell.name} />
        ))}
      </ul>
    </div>
  );
}

export default SpellsList;
