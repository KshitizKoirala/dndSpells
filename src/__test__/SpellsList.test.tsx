import { createRoot } from "react-dom/client";
import { render, screen } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { rest } from "msw";
import { setupServer } from "msw/node";

import SpellList from "../component/spells/SpellsList";
// import SpellSingle from "../component/spells/SpellSingle";
import { spellData } from "../mockData/spells";

/**
 * Making mock server to handle request
 */
const handlers = [
  rest.get("https://www.dnd5eapi.co/api/spells", (req, res, ctx) =>
    res(ctx.json(spellData), ctx.delay(150)),
  ),
];

const server = setupServer(...handlers);

beforeAll(() => server.listen());

afterEach(() => server.resetHandlers());

afterAll(() => server.close());

describe("home", () => {
  function prepare() {
    const { container } = render(<SpellList />);
    return { container };
  }

  it("testing match snapshot", () => {
    const { container } = prepare();
    expect(container).toMatchSnapshot();
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    const root = createRoot(div);
    // eslint-disable-next-line testing-library/no-unnecessary-act
    act(() => {
      root.render(<SpellList />);
    });
  });

  it("renders all the spells in the document", async () => {
    render(<SpellList />);
    expect(
      screen.getByRole("link", { name: spellData.results[0].name }),
    ).toBeInTheDocument();
  });
});
