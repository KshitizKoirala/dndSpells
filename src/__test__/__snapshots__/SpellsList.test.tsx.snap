// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home testing match snapshot 1`] = `
<div>
  <div
    class="spells-parent"
    data-testid="spells-list"
  >
    <ul
      class="mt-3 spells-list"
    />
  </div>
</div>
`;
